cartography/intel/aws/lambda_function.py

```diff
@@ -29,47 +33,193 @@ def get_lambda_data(boto3_session: boto3.session.Session, region: str) -> List[D
-    MERGE (lambda:AWSLambda{id: {Arn}})
+    UNWIND {lambda_functions_list} AS lf
+    MERGE (lambda:AWSLambda{id: lf.FunctionArn})
     ON CREATE SET lambda.firstseen = timestamp()
-    SET lambda.name = {LambdaName},
-    lambda.modifieddate = {LastModified},
-    lambda.arn = {Arn},
-    lambda.runtime = {Runtime},
-    lambda.description = {Description},
-    lambda.timeout = {Timeout},
-    lambda.memory = {MemorySize},
+    SET lambda.name = lf.FunctionName,
+    lambda.modifieddate = lf.LastModified,
+    lambda.runtime = lf.Runtime,
+    lambda.description = lf.Description,
+    lambda.timeout = lf.Timeout,
+    lambda.memory = lf.MemorySize,
+    lambda.codesize = lf.CodeSize,
+    lambda.handler = lf.Handler,
+    lambda.version = lf.Version,
+    lambda.tracingconfigmode = lf.TracingConfig.Mode,
+    lambda.revisionid = lf.RevisionId,
+    lambda.state = lf.State,
+    lambda.statereason = lf.StateReason,
+    lambda.statereasoncode = lf.StateReasonCode,
+    lambda.lastupdatestatus = lf.LastUpdateStatus,
+    lambda.lastupdatestatusreason = lf.LastUpdateStatusReason,
+    lambda.lastupdatestatusreasoncode = lf.LastUpdateStatusReasonCode,
+    lambda.packagetype = lf.PackageType,
+    lambda.signingprofileversionarn = lf.SigningProfileVersionArn,
+    lambda.signingjobarn = lf.SigningJobArn,
     lambda.lastupdated = {aws_update_tag}
-    WITH lambda
+    WITH lambda, lf
     MATCH (owner:AWSAccount{id: {AWS_ACCOUNT_ID}})
     MERGE (owner)-[r:RESOURCE]->(lambda)
     ON CREATE SET r.firstseen = timestamp()
     SET r.lastupdated = {aws_update_tag}
-    WITH lambda
-    MATCH (role:AWSPrincipal{arn: {Role}})
+    WITH lambda, lf
+    MATCH (role:AWSPrincipal{arn: lf.Role})
     MERGE (lambda)-[r:STS_ASSUME_ROLE_ALLOW]->(role)
     ON CREATE SET r.firstseen = timestamp()
     SET r.lastupdated = {aws_update_tag}
     """
 
+    neo4j_session.run(
+        ingest_lambda_functions,
+        lambda_functions_list=data,
+        Region=region,
+        AWS_ACCOUNT_ID=current_aws_account_id,
+        aws_update_tag=aws_update_tag,
+    )
+
+
+@timeit
+def get_function_aliases(lambda_function: Dict, client: botocore.client.BaseClient) -> List[Any]:
+    aliases: List[Any] = []
+    paginator = client.get_paginator('list_aliases')
+    for page in paginator.paginate(FunctionName=lambda_function['FunctionName']):
+        aliases.extend(page['Aliases'])
+
+    return aliases
+
+
+@timeit
+def get_event_source_mappings(lambda_function: Dict, client: botocore.client.BaseClient) -> List[Any]:
+    event_source_mappings: List[Any] = []
+    paginator = client.get_paginator('list_event_source_mappings')
+    for page in paginator.paginate(FunctionName=lambda_function['FunctionName']):
+        event_source_mappings.extend(page['EventSourceMappings'])
+
+    return event_source_mappings
+
+
+@timeit
+def get_lambda_function_details(
+        boto3_session: boto3.session.Session, data: List[Dict], region: str,
+) -> Generator[Any, Any, Any]:
```

Comment:
looks like this generator returns 4 things? also can you be more specific than just 'Any' here?

Reply:
Yeah. Made the change in the latest commit

Reply:
@kunaals So I reverted to the 3 'Any's because for some reason it fails in the linter - `error: "Generator" expects 3 type arguments, but 4 given`. I think this is why in all other codes too irrespective of how many values Generator returns - only 3 have always been mentioned.

Linter error Link - https://github.com/lyft/cartography/pull/623/checks?check_run_id=2691387616

Reply:
Saw this on Stackoverflow: https://stackoverflow.com/a/42531262/565573

> The return type of generator functions can be annotated by the generic type Generator[yield_type, send_type, return_type] provided by typing.py module

So I think you need to do

```
-> Generator[Tuple[Type1, Type2, Type3, Type4], None, None]
```

Admittedly I have no idea what a `send_type` or `return_type` is in this context, but that's the expectation for the `Generator` type.

Reply:
See https://stackoverflow.com/a/42531262/565573 - the type args for a generator are strange. See if you can make sense of that. If not, I won't block if typehints are not 100% correct here.

Reply:
Okay I took a look at that. 

The `yield_type` will remain `Any` since I have two types in yield - str (for lambda_function['FunctionArn']) and List[Dict] (for the rest of them).

I'm not sure what `send_type` is. But what I can glean from the context is it is the value that we pass to another function within this function. I feel for our context it can be `Any` since we are passing different types to different functions in `get_lambda_function_details()`.

Finally, `return_type` is if we have a return statement. Since we don't it is `None`. So I'm turning the Generator to [Any, Any, None].